services/src/main/java/org/keycloak/locale/DefaultLocaleUpdaterProvider.java

```diff
@@ -16,32 +16,51 @@
  */
 package org.keycloak.locale;
 
+import jakarta.ws.rs.core.UriInfo;
 import org.jboss.logging.Logger;
+import org.keycloak.events.Details;
+import org.keycloak.events.EventBuilder;
+import org.keycloak.events.EventType;
 import org.keycloak.models.KeycloakSession;
 import org.keycloak.models.RealmModel;
 import org.keycloak.models.UserModel;
 import org.keycloak.services.managers.AuthenticationManager;
 import org.keycloak.services.util.CookieHelper;
 import org.keycloak.storage.ReadOnlyException;
-
-import jakarta.ws.rs.core.UriInfo;
+import org.keycloak.userprofile.AttributeChangeListener;
+import org.keycloak.userprofile.UserProfile;
+import org.keycloak.userprofile.UserProfileContext;
+import org.keycloak.userprofile.UserProfileProvider;
 
 public class DefaultLocaleUpdaterProvider implements LocaleUpdaterProvider {
 
     private static final Logger logger = Logger.getLogger(LocaleSelectorProvider.class);
 
-    private KeycloakSession session;
+    private final KeycloakSession session;
 
     public DefaultLocaleUpdaterProvider(KeycloakSession session) {
         this.session = session;
     }
 
     @Override
     public void updateUsersLocale(UserModel user, String locale) {
-        if (!locale.equals(user.getFirstAttribute("locale"))) {
+        final String previousLocale = user.getFirstAttribute("locale");
+        if (!locale.equals(previousLocale)) {
             try {
+                EventBuilder event = new EventBuilder(session.getContext().getRealm(), session, session.getContext().getConnection())
+                        .event(EventType.UPDATE_PROFILE)
+                        .user(user)
+                        .client(session.getContext().getClient())
+                        .detail(Details.PREF_PREVIOUS + UserModel.LOCALE, previousLocale)
```

Comment:
I wonder about why we cannot rely on `EventAuditingAttributeChangeListener` to set those details automatically like other places calling `profile.update` method? Also about why we still need to manually call `user.setSingleAttribute` from `DefaultLocaleUpdaterProvider` instead of relying on profile to set it automatically.

Is it because of "locale" attribute is not automatically available in the `UPDATE_PROFILE` context? I wonder if we should have some dedicated context like `UPDATE_PROFILE_LOCALE`, which can be created manually and support only "locale" attribute? Or is it too "heavy" just for this purpose?
@pedroigor WDYT?

Reply:
Yeah, I think it does not make much sense to set the attribute directly if also using the user profile provider to update the user.

I'm not sure about creating a dedicated context but perhaps add locale to the update profile context, if possible.

Reply:
@mposolda @pedroigor But if I add locale to the UPDATE_PROFILE it would appear in the update profile page, right? Do we want that?

And thinking about this, if there is some invalid data in the user the locale update is going to fail (for example if a required attribute is not filled in the user). Maybe I was wrong and better don't use the user profile, just update the "locale" as before and trigger the event.

Reply:
+1 to avoid using user profile. As you pointed, it can be a risk of "required" attribute not being filled, which would cause that it won't be possible to update locale.

We have follow-up tasks to better handling various related situations. Perhaps we can have some more "dynamicity" in the user profile context to being able to say "Please validate just `locale` attribute now and ignore all other attributes or required attributes" etc...

So I would say just ignore userProfile for now and "hardcode" the details in the event like you did. Is it ok to update PR in this way?